Code/GraphMol/Wrap/rdchem.cpp

```diff
@@ -116,14 +110,111 @@ void WrapLogs() {
   if (rdWarningLog != nullptr) rdWarningLog->SetTee(warning);
 }
 
+python::tuple getAtomIndicesHelper(const KekulizeException &self) {
+  python::list res;
+  for (auto idx : self.getAtomIndices()) {
+    res.append(idx);
+  }
+  return python::tuple(res);
+}
+
+PyObject *molSanitizeExceptionType = nullptr;
+PyObject *atomSanitizeExceptionType = nullptr;
+PyObject *atomValenceExceptionType = nullptr;
+PyObject *atomKekulizeExceptionType = nullptr;
+PyObject *kekulizeExceptionType = nullptr;
+
+// pattern from here:
+// https://stackoverflow.com/questions/11448735/boostpython-export-custom-exception-and-inherit-from-pythons-exception
+template <typename EXC_TYPE>
+void sanitExceptionTranslator(const EXC_TYPE &x, PyObject *pyExcType) {
+  PRECONDITION(pyExcType != nullptr, "global type not initialized");
+  python::object pyExcInstance(python::handle<>(python::borrowed(pyExcType)));
+  pyExcInstance.attr("cause") = x;
+  PyErr_SetString(pyExcType, x.message());
+}
+
+// pattern from here:
+// https://stackoverflow.com/questions/9620268/boost-python-custom-exception-class
+PyObject *createExceptionClass(const char *name,
+                               PyObject *baseTypeObj = PyExc_ValueError) {
+  std::string scopeName =
+      python::extract<std::string>(python::scope().attr("__name__"));
+  std::string qualifiedName0 = scopeName + "." + name;
+  char *qualifiedName1 = const_cast<char *>(qualifiedName0.c_str());
+
+  PyObject *typeObj = PyErr_NewException(qualifiedName1, baseTypeObj, 0);
+  if (!typeObj) python::throw_error_already_set();
+  python::scope().attr(name) = python::handle<>(python::borrowed(typeObj));
+  return typeObj;
+}
+
 BOOST_PYTHON_MODULE(rdchem) {
   python::scope().attr("__doc__") =
       "Module containing the core chemistry functionality of the RDKit";
   RegisterListConverter<RDKit::Atom *>();
   RegisterListConverter<RDKit::Bond *>();
   rdkit_import_array();
+
+  // this is one of those parts where I think I wish that I knew how to do
+  // template meta-programming
+  python::class_<MolSanitizeException>("_cppMolSanitizeException",
```

Comment:
The main reason is that it’s important to make sure that MolSanitizeException (and it’s children) derive from ValueError on the Python side; otherwise we break existing code.

Reply:
The main reason is that it’s important to make sure that MolSanitizeException (and it’s children) derive from ValueError on the Python side; otherwise we break existing code.